### ISSUE TITLE
[Explainability Evaluation] - Fidelity +/- metrics

### ISSUE THREAD

**Issue by BlazStojanovic** [pyg-team/pytorch_geometric#5958]
### 🚀 The feature, motivation and pitch

Implement Fidelity +/- metrics (Fid+ and Fid-) for the purpose of evaluation explainability. This feature request is a part of broader explainability evaluation metrics (parent issue #5628) which will be crucial for the new revamp of explainability (see roadmap https://github.com/pyg-team/pytorch_geometric/issues/5520).

### Brief overview of Fidelity+/-
Fidelity measures check the explanations for their *faithfulness* to the model. This means that checking that that important inputs (nodes/edges/node features) according to the explanation are discriminative to the model. 

Formally fidelity is given with equations (from A.6 of [3])
![image](https://user-images.githubusercontent.com/22046931/201399747-4606da07-a8e3-4332-a72b-90590204f771.png)
For details and explanations see references, particularly [2-3].

### Implementation checklist
- Implement fidelity +/- for:
  - Phenomenon focus (measured with respect to ground truth labels)
  - Model focus (measured with respect to GNN outcome)
- Implement fidelity +/- for:
  - Probabilities
  - Accuracy
- Implement tests
- Implement examples

### References
1. [Explainability Methods for Graph Convolutional Neural Networks](https://openaccess.thecvf.com/content_CVPR_2019/papers/Pope_Explainability_Methods_for_Graph_Convolutional_Neural_Networks_CVPR_2019_paper.pdf)
2. [Explainability in Graph Neural Networks: A Taxonomic Survey](https://arxiv.org/abs/2012.15445)
3. [GraphFramEx: Towards Systematic Evaluation of Explainability Methods for Graph Neural Networks
](https://arxiv.org/abs/2206.09677)

**Comment by AbdullahWasTaken** [comment]
`ExplainerAlgorithm` can produce `node_mask`, `node_feat_mask`, and `edge_mask`. But in case one of these is missing, what should be the default behavior? Should the mask contain all ones or all zeros?

**Comment by BlazStojanovic** [comment]
@AbdullahWasTaken We’ve implemented some additional functionallity to `Explanation`, you can now use:
- `Explanation.get_explanation_subgraph()`, and
- `Explanation.get_complement_subgraph()`
to obtain the subgraphs you need!

**Comment by CXX1113** [comment]
In fact, I found that this formula in the GraphFramEx paper is wrong. S should have the subscript i (that is, S_i), instead of using the same mask for the prediction of each i. In the code of GraphFramEx, S is actually regarded as having S_i.
The code link is follow:
[https://github.com/DS3Lab/GraphFramEx/blob/main/code/evaluate/fidelity.py#L69C1-L73C42](url)

**Comment by BlazStojanovic** [comment]
> 

Not absolutely sure what you mean with $S_i$ but the explanations are of course specific to each node, i.e. their explanations differ and so do the respective subgraphs we use to evaluate fidelity. In this case you are right that $S$ is actually regarded as $S_i$, and we treat it as such in PyG: https://github.com/pyg-team/pytorch_geometric/blob/40d4718713142bd60a9f65e4d66a80789290ae84/torch_geometric/explain/metric/fidelity.py#L75C4-L75C4

As for the notation in the paper, I think the authors just wanted to avoid clutter and not use index twice, 

![image](https://github.com/pyg-team/pytorch_geometric/assets/22046931/9ea807cc-99e6-47c4-81eb-f3c2d1640978)

I think the definition from the paper indicates quite clearly that the masked graphs (and complements) are node specific.
